Using Edit Distance to Word-Diff Across Multiple Lines

yokurang opened the pull request.

yokurang committed:
initial commit

yokurang committed:
saving progress

yokurang committed:
refactored tests to match pair lines implementation

yokurang committed:
some progress in pair lines

panglesd commented on lib/WordDiff.ml:
```diff
@@ -43,98 +77,67 @@ let edit_distance (type a) (compare : a -> a -> bool) (s : a array)
+        construct_pairs [] ys lcs ((None, Some y) :: acc)
     | x :: xs, y :: ys, [] ->
-        construct_diff xs ys [] (Changed x :: acc_mine) (Changed y :: acc_their)
+        construct_pairs xs ys [] ((Some x, Some y) :: acc)
```
Pairing these lines is wrong!

yokurang replied:
Hi, do you mind elaborating more on why it is wrong?

panglesd replied:
Try it on the following test!
```diff
- aaaaaaaaa a
- egdg xsd fi sdgfyu s gd csue c e yeg scy
+ aaaaaaaaa b
+ yugyu fft yu utf ytfutf uyg gygu yftuf gg uf ft
```

yokurang replied:
LACS -> A sequence of lines 
Here, we have two lines in 1 and two lines in 2. 
So LCAS is a matter of lines. Here, it will be a length of 1, and the LCAS are (Some aaaaaaaaa a, Some aaaaaaaaa b)

So here, in the accumulator, I am adding x and y as being paired, when x and y are not part of the LACS. 
Here, the example, x and y corresponds to egdg ... and yugyu ...
So they should not be paired.

yokurang committed:
pushing latest progress

panglesd committed:
Pair programming session

- Fixed bugs in the pairing of lines
    The LACS must consist of the _two_ subsequence
- Finished rendering paired lines

Co-authored-by: Alan Matthew <yokurang@gmail.com>
